[Test] Creating TC_CADMIN_1_3 and TC_CADMIN_1_4 python test modules

j-ororke opened the pull request:
- Current WIP python script creation for TC_CADMIN_1_3 following test steps in PR: [4689](https://github.com/CHIP-Specifications/chip-test-plans/pull/4689)
- Added WIP python script for TC_CADMIN_1_4 following test steps in PR: [4730](https://github.com/CHIP-Specifications/chip-test-plans/pull/4730)

#### Testing

j-ororke committed:
[Test] Creating TC_CADMIN_1_3 python test module:
- Current WIP script creation following test steps in PR: https://github.com/CHIP-Specifications/chip-test-plans/pull/4689

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

arkq commented on src/python_testing/TC_CADMIN_1_3.py:
```diff
@@ -0,0 +1,249 @@
+#    limitations under the License.
+#
+# === BEGIN CI TEST ARGUMENTS ===
+# test-runner-runs: run1
```
Could you use YAML format for CI args?
See this for reference: https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/TC_ACE_1_2.py

I'm planing to remove support for not-YAML format in a future, since it does not support line-breaks.

j-ororke replied:
Thank you for letting me know about this @arkq, 
I have updated the CI arguments to using YAML format in this test script now. 
I will start doing this for the other CADMIN tests that are currently being worked on.

j-ororke committed:
Update TC_CADMIN_1_3:
- Updated TC_CADMIN_1_3 to latest CI format

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Updated TC_CADMIN_1_3:
- Resolving linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Renamed TC_CADMIN_1_3 to TC_CADMIN_1_3_4 test module:
- Appended WIP CADMIN_1_4 test script to TC_CADMIN_1_3 test module
- Renamed TC_CADMIN_1_3 to TC_CADMIN_1_3_4
- Added issue in a TODO comment since current workaround in place to make CADMIN_1_3 pass

j-ororke committed:
Update TC_CADMIN_1_3_4.py

- Updated test step 14 for TC_CADMIN_1_3 to contain correct information

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Minor change to attempt resolving restylizer issue

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Updated TC_CADMIN_1_3_4 test module:
- Updating to resolve some linting issues

arkq commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,378 @@
+# test-runner-runs:
+#   run1:
+#     app: ${ALL_CLUSTERS_APP}
+#     factoryreset: true
```
```suggestion
#     factory-reset: true
```

There is a small change like that in current master (to avoid spellcheck warning)

j-ororke replied:
Thank you, I will update this to reflect that change!

j-ororke committed:
Update src/python_testing/TC_CADMIN_1_3_4.py

Updating to reflect change in master for "factory-reset", thank you @arkq

Co-authored-by: Arkadiusz Bokowy <arkadiusz.bokowy@gmail.com>

j-ororke committed:
Updated TC_CADMIN_1_3_4 test module:
- Now using get_setup_payload_info() to gather discriminator and passcode for BCM commissioning in TC_CADMIN_1_4

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Resolved linting issue

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Updating TC_CADMIN_1_3_4 and matter_testing modules:
- Updating TC_CADMIN_1_3_4 due to change to matter_testing_support module
- Updating matter_testing module to resolve missing dependencies not found in chip library.

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Update matter_testing.py

Reverting changes back as just needed to rebuild the python environment to get this working.

j-ororke committed:
Update matter_testing.py

Updating location of imported new modules for chip.testing to match with master

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Updated TC_CADMIN_1_3_4 test module:
- Added revoke commissioning after test step 13 to revoke commissioning window.

j-ororke committed:
Updated TC_CADMIN_1_3_4 test module:
- Enabled verbose output to identify issue noticed in CI/CD pipeline Linux REPL tests

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,389 @@
+        th2_certificate_authority = self.certificate_authority_manager.NewCertificateAuthority()
+        th2_fabric_admin = th2_certificate_authority.NewFabricAdmin(vendorId=0xFFF1, fabricId=self.th1.fabricId + 1)
+        self.th2 = th2_fabric_admin.NewController(nodeId=2)
+        setupPayloadInfo = self.get_setup_payload_info()
```
So this is a bit of a dragon that won't be immediately apparent in the CI. get_setup_payload_info won't necessarily have anything populated if the device is already commissioned when this test starts. So you'll have to enforce that the tester supplies the setup_payload in some form even if the device is already commissioned (ie test run with no factory reset). If you could make this error and the explanation of why this is required VERY explicit, it will be helpful at cert. So two suggestions here:
1) move this up to the top, verify that you have something returned and if not assert with a very explicit message that the setup payload is required for this test even if the device is already commissioned because it is testing the basic commissioning window, which uses the original passcode and discriminator
2) add an extra run of this test to the CI that doesn't use factory reset, to be run right after the original run to verify this works with a pre-commissioned device.

j-ororke replied:
Hi Cecille,
Thank you for the suggestions and tips on this situation above, I have made the changes to the script to hopefully resolve the above comments suggestions, please let me know if this works to what you were expecting and if you think it will be acceptable.

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Removed local OpenCommissiongWindow and CommissionAttempt functions
- Started using openCommissioningWindow function from matter_testing support module

restyled-commits committed:
Restyled by autopep8

j-ororke commented on src/python_testing/TC_CADMIN_1_3.py:
```diff
@@ -0,0 +1,255 @@
+
+
+class TC_CADMIN_1_3(MatterBaseTest):
+    async def CommissionAttempt(
```
I have removed this function from the code and started using CommissionOnNetwork in the code directly as you mentioned. Thank you it is a bit cleaner now than before!

j-ororke commented on src/python_testing/TC_CADMIN_1_3.py:
```diff
@@ -0,0 +1,255 @@
+        window_status = await self.read_single_attribute_check_success(dev_ctrl=self.th2, fabric_filtered=True, cluster=AC_cluster, attribute=AC_cluster.Attributes.WindowStatus)
+        return window_status
+
+    async def OpenCommissioningWindow(self, th: ChipDeviceCtrl, duration: int) -> CommissioningParameters:
```
Hi Cecille, 
I have replaced the above local function in this test module and started using the provided function from matter_testing helper module in its place.

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Resolving lint errors

j-ororke committed:
Added TC_CADMIN_1_4_nofreset test module:
- Added new test module to verify that if no factory reset occurs between TC_CADMIN_1_3_4 test modules and TC_CADMIN_1_4_nofreset, how to handle that situation for BCM.

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated TC_CADMIN_1_4_nofreset test module:
- Resolved linting errors

j-ororke committed:
Updated TC_CADMIN_1_4_nofreset test module and manualTests.json:
- Removed TC_CADMIN_1_3 and TC_CADMIN_1_4 from manualTests.json MultipleFabrics section
- Updated TC_CADMIN_1_4_nofreset to remove commissioning attempt before start of test.

j-ororke committed:
Moved TC_CADMIN_1_4_nofreset to unit testing framework:
- Moved TC_CADMIN_1_4_nofreset to unit testing framework folder and renamed it to test_TC_CADMIN_1_4

j-ororke committed:
Updated TC_CADMIN_1_3_4 and unit test test_TC_CADMIN_1_4
- Placed assert for if get_setup_payload_info() at start of test_CADMIN_1_4 test function
- updated test_TC_CADMIN_1_4 test function name

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updated test_TC_CADMIN_1_4:
- Resolving linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated tests.yaml:
- Added test_CADMIN_1_4 python unit-test test module to attempt running it in CI

j-ororke committed:
Updated test_TC_CADMIN_1_4 unit-test module:

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updated test_TC_CADMIN_1_4 test module

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating test_TC_CADMIN_1_4 unit-test module

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updated test_TC_CADMIN_1_4 unit-test module:
- Attempting to get import to be imported so test runs without restyler changing it back

j-ororke committed:
Updating test_TC_CADMIN_1_4 unit-test module:
- Resolving lint error

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Updating CI argument to turn on quiet mode for this test module

j-ororke committed:
Updated test_TC_CADMIN_1_4:
- Updated to using isort off to avoid CI re-sorting the dependencies during this test run

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating test_TC_CADMIN_1_4 unit-test module:
- Updating to get isort to not sort dependicies for this test

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating test_TC_CADMIN_1_4 unit-test module:
- Reverting back to setting dependency in test function as not able to get isort to not sort it out of position

restyled-commits committed:
Restyled by isort

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Added TC_CADMIN_1_4_noreset test module and updated test_TC_CADMIN_1_4:
- Added TC_CADMIN_1_4_noreset test module to attempt validation of script with no factory reset between tests
- Updated test_TC_CADMIN_1_4 unit-test module to exit if credentials not found.

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updated TC_CADMIN_1_4_noreset test module:
- Resolving lint issues

j-ororke committed:
Updated TC_CADMIN_1_4_noreset test module:
- reorganizing CI args to attempt testing with factory-reset disabled

j-ororke committed:
Updating TC_CADMIN_1_4_noreset, tests.yaml, and execute_python_test:
- Updating to attempt to get TC_CADMIN_1_4_noreset to get it to run in CI

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated TC_CADMIN_1_4_noreset test module:
- Removed commissioning-method CI argument to not run commissioning before start of this test

j-ororke committed:
Delete src/python_testing/test_testing/test_TC_CADMIN_1_4.py

Removing unneeded small unit-test due to not needed

j-ororke committed:
Update tests.yaml

Removed prior unit-test test_TC_CADMIN_1_4.py as it was deleted

j-ororke committed:
Removed Test_TC_CADMIN_1_3 and Test_TC_CADMIN_1_4 yaml's:
- These 2 yaml scripts are no longer needed as they are being replaced by the python test modules.

j-ororke committed:
Updated TC_CADMIN_1_3_4 and TC_CADMIN_1_4_noreset:
- Disabled verbose output in CI arguments (quiet mode)

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,365 @@
+        current_fabric_index = await self.read_single_attribute_check_success(dev_ctrl=th, endpoint=0, cluster=cluster, attribute=attribute)
+        return current_fabric_index
+
+    def pics_TC_CADMIN_1_3(self) -> list[str]:
```
can these two tests be combined into a single parameterized function?

cecille replied:
Looking at these two tests, from step 1 to step 7, the steps are basically identical with the exception of the command sent (open commissioning window vs. open basic commissioning window. Can definitely re-use some of this between tests.

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,365 @@
+            fabric_info = await self.read_single_attribute_check_success(dev_ctrl=th, fabric_filtered=True, cluster=OC_cluster, attribute=OC_cluster.Attributes.Fabrics)
+        return fabric_info
+
+    async def get_rcac_decoded(self, th: str) -> int:
```
is sending the csr request here just so you can issue a noc chain so you can get the rcac? If so, you shouldn't need to jump through so many hoops - the CA manager is

self.certificate_authority_manager and it maintains a list of the active roots: activeCaList

j-ororke replied:
I have resolved the above comment, I was able to use the self.certificate_authority_manager.activeCaList method as you mentioned above in the test module and was able to remove the function that was being used previously.

One caveat, however, since running the test 2 iterations in CI without factory-reset inbetween. I have added a try and except clause to TC_CADMIN_1_4 test to make sure that it uses the correct root public key from self.certificate_authority_manager.activeCaList during the second iteration.

j-ororke committed:
Updating TC_CADMIN_1_3_4 and tests.yaml:
- Removed calls in tests.yaml for TC_CADMIN_1_3_4 as not needed individual calls for tests now that we have changed the CI arguments
- Removed TC_CADMIN_1_4_noreset standalone test module as we are now using the CI arguments
- Removed the line from execute_python_tests test runner module to remove not call for TC_CADMIN_1_4_noreset test module, as no longer needed
- Renamed class for tests to TC_CADMIN from TC_CADMIN_1_3_4 to possibly encompaculate all CADMIN tests eventually.
- Changed method to get th2_fabric_info to match with the method used to get th1_fabric_info in TC_CADMIN_1_3_4 test module

restyled-commits committed:
Restyled by autopep8

j-ororke commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,365 @@
+    async def get_fabrics(self, th: ChipDeviceCtrl) -> int:
+        OC_cluster = Clusters.OperationalCredentials
+        if th == self.th2:
+            th2_fabric_info = await th.ReadAttribute(nodeid=self.dut_node_id, fabricFiltered=True, attributes=[(0, OC_cluster.Attributes.Fabrics)])
```
Hi Cecille,
Sorry, I believe there was a reason I did this previously. 
However, it appears that it does now work the same as TH1's method for getting the fabric information. 
So, I have changed TH2's method to now match with TH1's method for getting the fabric information.

j-ororke commented on src/python_testing/TC_CADMIN_1_4_noreset.py:
```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env -S python3 -B
```
Thank you, 
I was not aware that I could do that in this manner, so I have changed the CI arguments as you have mentioned and removed the prior TC_CADMIN_1_4_noreset test module as it will no longer be needed.

j-ororke committed:
Updated TC_CADMIN_1_3_4 test module:
- Updated method to cross verify root public key, now using self.certificate_authority_manager.activeCaList

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Disabling quiet mode in CI arguments

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Resolving linting errors with unused imports

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Updating to resolve linting issue

j-ororke committed:
RCAC data callback implementation during commeissioning:
- Updated OpCredsBinding C++ module to include callback functionality to gather RCAC in chip cert format during kGenerateNOCChain commissioning stage
- Updated ChipDeviceCtrl python module to include new functions to establish callback during OnNetworkCommissioning and added a new arg "" for this functionality to trigger the callback when wanted to gather rcac data, also passing that data into tests based off new arg value
- Updated TC_CADMIN_1_3_4 test module to implement gathering and using the RCAC data to validate the rootpublickey matches during testing.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated OpCredsBinding C++ module:
- Removed some unneccessary dependencies that had been used during debugging

j-ororke committed:
Updating OpsCredsBinding.cpp and ChipDeviceCtrl.py:
- Some further cleanup performed.

restyled-commits committed:
Restyled by clang-format

cecille commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -687,4 +748,29 @@ PyChipError pychip_GetCompletionError()
+        }
+        else
+        {
+            ChipLogError(Controller, "Python RCAC callback is not set.");
```
You can omit this else - gPythonRCACCallback was set in this same function. If it was set to nullptr, it was intentional.

j-ororke replied:
Thank you, I will remove this from the code!

cecille commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -687,4 +748,29 @@ PyChipError pychip_GetCompletionError()
 
+extern "C" {
+// Define a global callback to set the RCAC data callback
+using PythonRCACCallback                      = void (*)(const uint8_t *, size_t);
```
General question - why do this as a callback and a future rather than just a function to get a blob of data that's stored in the test commissioner? The RCAC isn't really valid until the commissioning completes anyway, so is there a point of separating the callbacks?

cecille replied:
So here's the concern with this - the commissioning delegate is already quite complex and you're introducing a fair amount of new complexity here to just surface the RCAC byte values at the point when they're captured. But there's no reason for any of the tests to attempt to access the RCAC until after the commissioning completes anyway since it's not valid until you know the device has accepted it. 

Does the commissioning process hang if you do nothing but copy out the RCAC to the vector? It shouldn't. I'm concerned what's actually happening here is a premature completion signal based on this new future.

basically, this SHOULD work with just a copy to the vector and none of this callback stuff. If you can get it working at that point, all this callback stuff can be removed and you should be able to retrieve the data after the commissioning completes.

j-ororke replied:
Thank you for bringing this up, I had not considered what you have mentioned above previously. I completely agree with you, it is always best to keep the code as simple as possible while completing its purpose as long as it's not causing any additional problems. In this point the callbacks that were previously established were adding complexity and possible issues that were going unnoticed during my testing unknowingly at the time.

When I first attempted implementing the code the hang appeared to have been due to attempting to implement the pulling point of the RCAC/RPK from the OnCommissioningComplete function in the OnNetworkCommissioning function of ChipDeviceCtrl python module. However, since moving that outside the embedded OnNetworkCommissioning function and making sure that the new functionality was implemented correctly, while also providing the expected value back, then we noticed no further hangs or issues when running the new code.

Updated the new code to remove the callbacks and replaced with a similar function that you have mentioned above, attempting to keep the code simple without adding any unnecessary complexity to it that was being caused by the callbacks, this should hopefully remove any underlying issues that could have gone previously unnoticed.

j-ororke commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -292,9 +326,35 @@ class TestCommissioner : public chip::Controller::AutoCommissioner
     CHIP_ERROR GetCompletionError() { return mCompletionError; }
 
+    // Getter method to access RCAC data
+    // const std::vector<uint8_t>& GetRCACData() const { return mRCACData; }
```
Thank you, 
Sorry forgot to remove some leftover debugging code, I have removed the above lines and the member var for mRCACData

j-ororke committed:
Updated OpCredsBinding C++ module:
- Removed some unneccessary code for cleanup of debugging session
- Removed else statement from pychip_SetCommissioningRCACCallback void function

j-ororke commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -292,9 +326,35 @@ class TestCommissioner : public chip::Controller::AutoCommissioner
         }
     }
 
+    CHIP_ERROR GetCommissioningResultError() const { return mCommissioningResultError; }
```
Thank you, 
This has been removed from the code.

j-ororke committed:
Updated OpCredsBinding C++ module:
- Changed from using ScopedMemoryBuffer to establishing pointer in vector to avoid leaks

j-ororke commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -149,6 +152,35 @@ class TestCommissioner : public chip::Controller::AutoCommissioner
+                // Converting rcac to chip cert format to be decyphered by TLV later in python3
+                chip::Platform::ScopedMemoryBuffer<uint8_t> chipRcac;
+                MutableByteSpan chipRcacSpan;
+                chipRcac.Alloc(Credentials::kMaxCHIPCertLength);
```
My apologies so I appear to have been mistaken here. Thank you @cecille for delving into the ScopedMemoryBuffer further as you found and let me know that it appears to not auto-free on destruction. Thus, it could indeed have caused leaks the way I had it coded.
I have changed the above code from using the ScopedMemoryBuffer to using a vector as you mentioned.

cecille commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -149,6 +152,32 @@ class TestCommissioner : public chip::Controller::AutoCommissioner
+                std::vector<uint8_t> chipRcac(Credentials::kMaxCHIPCertLength);
+                MutableByteSpan chipRcacSpan(chipRcac.data(), chipRcac.size());
+                chip::Credentials::ConvertX509CertToChipCert(rcacByteSpan, chipRcacSpan);
+                mCHIPRCACData.assign(chipRcacSpan.data(), chipRcacSpan.data() + chipRcacSpan.size());
```
can you just do mCHIPRCACDATA = chipRcac?

j-ororke committed:
Updating OpCredsBinding C++ module and ChipDeviceCtrl python module:
- Removed prior callbacks and replaced with function to remove additional complexity that could cause issues, making it simple

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

j-ororke committed:
Update manualTests.json

Removing files TC_CADMIN_1_21 and TC_CADMIN_1_22 as latest merge from TC_CADMIN_1_22 and TC_CADMIN_1_24 should have removed those files from this file.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

cecille commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -149,6 +151,33 @@ class TestCommissioner : public chip::Controller::AutoCommissioner
+            {
+
+                auto nocChain = report.Get<chip::Controller::NocChain>().rcac;
+                MutableByteSpan rcacSpan(const_cast<uint8_t *>(nocChain.data()), nocChain.size());
```
Does this get used anywhere besides creating the non-mutable span below? Is it possible to get rid of this? Does it actually need to be mutable? i don't think the conversion is in-place.

j-ororke replied:
Thank you, you are right, I did not need to create the mutable span in the code to get this to work. 
I have updated the code to just converting the x509 cert data stored in nocChain to the chipRcacSpan mutable span now.

cecille commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2228,7 +2260,32 @@ async def CommissionOnNetwork(self, nodeId: int, setupPinCode: int,
             )
 
-            return await asyncio.futures.wrap_future(ctx.future)
+            # If RCAC data is needed, await the result
```
can this be pulled out into a separate function?

j-ororke replied:
I have moved setting the rcac_bytes variable to inside a new function get_rcac() to clean this up a little bit, it still gets returned in the OnNetworkCommission() here but a bit cleaner now.
Let me know if this will be acceptable.

cecille commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2228,7 +2260,32 @@ async def CommissionOnNetwork(self, nodeId: int, setupPinCode: int,
+            # If RCAC data is needed, await the result
+            if get_rcac:
+                try:
+                    sleep(60)
```
why is this required?

j-ororke replied:
I was able to remove this sleep after moving the return await asyncio.futures.wrap_future(ctx.future) to above where the rcac_bytes variable gets set in this CommissionOnNetwork().

j-ororke committed:
Updating ChipDeviceCtrl and TC_CADMIN_1_3_4 modules:
- Removed additional commented out line of code in TC_CADMIN_1_3_4 test module.
- Updated ChipDeviceCtrl module to move rcac_bytes var setting to a new function, removed additional print line.
- Updated ChipDeviceCtrl module to await before setting the rcac_bytes var.

j-ororke committed:
Updating OpCredsBinding C++ module:
- Removed rcacByteSpan var as not needed in order to convert data before passing data to python3

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating ChipDeviceCtrl module:
- Resolving linting error

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module:
- Changing to using self.th1.OpenCommissioningWindow() in place of self.openCommissioningWindow() to resolve issue noticed in CI.

restyled-commits committed:
Restyled by autopep8

cecille commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -687,4 +720,24 @@ PyChipError pychip_GetCompletionError()
 
+extern "C" {
+// Function to get the RCAC data from the sTestCommissioner
+void pychip_GetCommissioningRCACData(uint8_t ** rcacDataPtr, size_t * rcacSize)
```
if you do uint8_t * const * rcacDataPtr, can you get rid of the const cast below?

cecille replied:
Better still, can you have the python side allocate space and copy out? That way you're not handing over the c++ memory to python.

j-ororke replied:
Hi Cecille,
I have updated the code to attempt to the best of my knowledge and ability to update allocating the memory in ChipDeviceCtrl python module's get_rcac function, then using memcpy in C++ I have copied the memory from C++ into the allocated python space.
Please let me know if this will work to resolve this

cecille commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2228,7 +2259,36 @@ async def CommissionOnNetwork(self, nodeId: int, setupPinCode: int,
+            res = await asyncio.futures.wrap_future(ctx.future)
+
+            # If RCAC data is wanted, attempt to pull the result
+            if get_rcac:
```
can you get rid of this and just have the caller call get_rcac directly? I'd rather keep the API for commissioning the same across all the network transports as much as possible.

j-ororke replied:
I have moved the get_rcac() to a standalone function that can now be called from test modules, such as self.th2.get_rcac() once commissioning has been completed.

j-ororke committed:
Updating ChipDeviceCtrl and TC_CADMIN_1_3_4 modules:
- Moved get_rcac() to a standalone function that can be called after commissioning is completed in ChipDeviceCtrl module
- Updated TC_CADMIN_1_3_4 test module to using the new standalone function self.th.get_rcac()

j-ororke committed:
Updating OpCredsBinding C++ and ChipDeviceCtrl python modules:
- Changing to allocating memory from python to store RCAC data captured from C++ module

j-ororke committed:
Updated ChipDeviceCtrl python module:
- Changed buffer size for storing RCAC to 650 from 1024

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by autopep8

cecille commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2191,6 +2202,27 @@ def GetFabricCheckResult(self) -> int:
+    '''
+
+    # Function to await the RCAC data
+    async def get_commissioning_rcac_data_async(self):
```
why is this async?

cecille replied:
And, actually - is this used anywhere?

j-ororke replied:
Thank you, sorry this was leftover it appears from prior coding down to implement getting RCAC data back from C++ code.

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,376 @@
+        th1_fabric_info = await self.get_fabrics(th=self.th1)
+
+        # Verify that the RootPublicKey matches the root public key for TH_CR1 and the NodeID matches the node ID used when TH_CR1 commissioned the device.
+        await self.send_single_cmd(dev_ctrl=self.th1, node_id=self.dut_node_id, cmd=Clusters.GeneralCommissioning.Commands.ArmFailSafe(10))
```
why is the failsafe armed here?

j-ororke replied:
So, I believe the way we were getting the RCAC initially had required us to set the failsafe in order to interact with the endpoints and get the RCAC data back for the validation done in those test steps. 
However, now with the new method of getting the RCAC is in place these fail safes appear to be no longer needed. 

Thank you for pointing these fail safes are no longer needed, I will remove these from the code.

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,376 @@
+            asserts.fail("DUT node ID from fabric does not equal DUT node ID for TH1 during commissioning")
+
+        # Expiring the failsafe timer in an attempt to clean up before TH2 attempt.
+        await self.th1.SendCommand(self.dut_node_id, 0, Clusters.GeneralCommissioning.Commands.ArmFailSafe(0))
```
same Q about disarm - nothing in here should be communicating with the DUT, so what is the failsafe doing here?

j-ororke replied:
Thank you, you are correct, the fail safes were no longer needed with the new method of getting the RCAC back. 
The initial way had implemented caused need to implement the fail safes in these areas, but fail safes are no longer needed.

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,376 @@
+        th2_fabric_info = await self.get_fabrics(th=self.th2)
+
+        # Verify that the RootPublicKey matches the root public key for TH_CR2 and the NodeID matches the node ID used when TH_CR2 commissioned the device.
+        await self.send_single_cmd(dev_ctrl=self.th2, node_id=self.dut_node_id, cmd=Clusters.GeneralCommissioning.Commands.ArmFailSafe(10))
```
ditto

j-ororke replied:
Thank you, this has been removed

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,376 @@
+        if th2_fabric_info[0].nodeID != self.dut_node_id:
+            asserts.fail("DUT node ID from fabric does not equal DUT node ID for TH1 during commissioning")
+
+        await self.th2.SendCommand(self.dut_node_id, 0, Clusters.GeneralCommissioning.Commands.ArmFailSafe(0))
```
same - I think both of these can be removed.

cecille replied:
and below.

j-ororke replied:
Thank you, these have been removed as well now I believe

j-ororke commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -1959,6 +1963,13 @@ def _InitLib(self):
             self._dmLib.pychip_GetCompletionError.argtypes = []
             self._dmLib.pychip_GetCompletionError.restype = PyChipError
 
+            #self._dmLib.pychip_SetCommissioningRCACCallback.argtypes = [_RCACCallbackType]
```
Sorry, these commented out python bindings have been removed.

j-ororke committed:
Updated ChipDeviceCtrl and TC_CADMIN_1_3_4 python modules:
- Removed unneeded get_commissioning_rcac_data_async()
- Removed commented out python bindings
- Removed fail safes as no longer needed with new method of gathering RCAC in place

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated TC_CADMIN_1_3_4 test module:
- Created combined_commission_val_steps function in order to combine test steps 1-7 to be used for both CADMIN_1_3 and CADMIN_1_4 tests.

restyled-commits committed:
Restyled by autopep8

cecille commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -687,4 +720,25 @@ PyChipError pychip_GetCompletionError()
+    *rcacSize = rcacData.size();
+
+    // Copy the data from C++ to Python's allocated memory
+    std::memcpy(rcacDataPtr, rcacData.data(), *rcacSize);
```
Does the python caller pass in the size of the allocated buffer? ie, how do you know that the rcacDataPtr buffer is large enough for the data you're coping into it?

cecille replied:
I know it's not a guarantee, but if you have the caller pass in the buffer size then you can check before you memcpy. It's not foolproof, but at least you can say you tried. 

The API here is public, so you're not guaranteed to be the only user. If the memcpy is too large, you'll segfault.

j-ororke replied:
Ah ok, understood, sorry I did not realize the API here was shared. 
I have updated to passing in the buffer size to the C++ function now, I have it doing an allocation check to make sure the rcac_size does not overflow the buffer we created in python3, then updated the bindings and that appears to be working at this time for returning the RCAC data.

cecille replied:
can you log an error and return a zero sized buffer if the value sent in is too small instead? This will give back a partial certificate, which is not useful.

j-ororke replied:
Sorry, good point, I have updated the code here to include a check, in the event that the buffer is smaller than the RCAC data size it should now return a zero sized buffer if the check fails.

cecille commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2230,6 +2237,30 @@ async def CommissionOnNetwork(self, nodeId: int, setupPinCode: int,
+    def get_rcac(self):
+        # Passes captured RCAC data back to python test modules to be used for validation
+        try:
+            rcac_size = 650
```
Max size for TLV is 400. You can use that, but also please add a comment as to why the value is used. (6.1.3 in the spec)

j-ororke replied:
I will try using 400 as the value, I will make sure to get a comment up for this in the code for why that size is used.
Thank you for providing the reference to 6.1.3 in the spec for this!

cecille commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -0,0 +1,330 @@
+        fabric_info = await self.read_single_attribute_check_success(dev_ctrl=th, fabric_filtered=True, cluster=OC_cluster, attribute=OC_cluster.Attributes.Fabrics)
+        return fabric_info
+
+    async def get_txt_record(self):
```
would you mind just getting Raul to take a peek here and make sure this API is correct? It looks correct to me, but he wrote it so I want to get a second set of eyes on it.

j-ororke replied:
I can respectfully request Raul when he has some time to take a look at this section of the code to make sure it looks correct.

cecille reviewed (commented):
Looks good with a couple of minor comments. Just ping me back once those are done and I'll checkmark it.

j-ororke committed:
Merge branch 'master' into create_CADMIN_1_3_test

restyled-commits committed:
Restyled by clang-format

j-ororke committed:
Updating OpCredsBinding and ChipDeviceCtrl modules:
- Update to OpCredsBinding C++ module was to add a missing bracket
- Update ChipDeviceCtrl python module was to change to using 400 instead of 650 for buffer size to store RCAC data

j-ororke committed:
Updating OpCredsBinding and ChipDeviceCtrl modules:
- Including change to pass in set buffer size for allocation check before passing RCAC data back to python3 from C++ module

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Update ChipDeviceCtrl.py

Updated to include reference to key sizes in spec 6.1.5

cecille commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2252,6 +2260,37 @@ async def CommissionOnNetwork(self, nodeId: int, setupPinCode: int,
+                rcac_data = bytearray(rcac_buffer[:actual_rcac_size.value])
+                rcac_bytes = bytes(rcac_data)
+            else:
+                raise Exception("RCAC data is empty")
```
Q - why raise an exception then catch it and return None? Why not just log and return here?

j-ororke replied:
Sorry, I think I was only attempting to return a value due to needing something to be returned to the test script for validation. Not thinking about doing it without the try and except block, the code has been modified quite a bit from the original form, I believe the try and except block had included more code before. 

However, I do apologize you are correct I should updated this section of code to catch and return here without adding in the try and except block being utilized here now that we have been able to make this code much cleaner.

cecille commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2252,6 +2260,37 @@ async def CommissionOnNetwork(self, nodeId: int, setupPinCode: int,
+            else:
+                raise Exception("RCAC data is empty")
+
+        except Exception as e:
```
except Exception is too broad - If you're expecting a particular exception, catch that. If this is just the one you threw, remove the try: except: and just return from the else.

j-ororke replied:
Thank you, that does make this section of the code cleaner without the try and except block. I believe initially I had only thought about returning a value here.

j-ororke committed:
Updating OpCredsBinding and ChipDeviceCtrl modules:
- Removed try and except block from get_rcac()
- Changed method to make sure that we aren't getting returned a partial cert

j-ororke committed:
Updating ChipDeviceCtrl python module:
- Updating to include better logger message when RCAC returns 0 sized data
- Updating to remove prior commented out no longer needed code block

j-ororke committed:
Update OpCredsBinding C++ module

- Moved comment into correct indent area

j-ororke committed:
Merge remote-tracking branch 'upstream/master' into create_CADMIN_1_3_test

restyled-commits committed:
Restyled by clang-format

j-ororke committed:
Updating OpCredsBinding C++ module:
- Resolving linting errors for log message if RCAC data size greater than allocated buffer size

j-ororke committed:
Updating OpCredsBinding C++ module:
- Updating Logging message for allocated buffer being too small to house RCAC data size, this should hopefully not happen.

raul-marquez-csa commented:
LGTM

andy31415 commented:
Fast track - this has been pending with checkmars for a while now.
